File: gunkim/2024_Weekly_I_Solved/24_02/6236_용돈관리.java
import java.util.*;
import java.io.*;

public class Main {


    private int[] money;

    public void solution() throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        money = new int[n];
        int maxMoney = 0;
        int totalMoney = 0;
        for (int i = 0; i < n; i++) {
            money[i] = Integer.parseInt(br.readLine());
            maxMoney = Math.max(maxMoney, money[i]);
            totalMoney += money[i];
        }

        // left는 무조건 maxMoney 보다는 커야한다. 그래야 인출했을 때 돈을 사용할 수 있음
        int left = maxMoney;
        int right = totalMoney;
        int answer = 0;
        while (left <= right) {
            // 인출 금액
            int mid = (left + right) / 2;

            // 인출 금액을 꺼내 쓸텐데 m번 안에 꺼내서 쓸 수 있는지 확인
            if (m >= isPossible(mid)) {
                answer = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        System.out.println(answer);

    }

    public int isPossible(int withdrawalAmount) { // withdrawalAmount : 인출 금액
        int count = 1;
        int used = withdrawalAmount;

        for (int m : money) {
            // 기본적으로 남은 금액이 있다면 해당 금액에서 사용을 한다
            used -= m;

            // 그런데 돈이 모자라면 남은 금액은 통장에 다시 넣어야 함
            // 돈을 다시 인출하여 해당 금액에서 사용한다
            if (used < 0) {
                count++;
                used = withdrawalAmount - m;
            }
        }
        // 돈을 몇 번 인출하는지 리턴
        return count;
    }



    public static void main(String[] args) throws Exception {
        new Main().solution();
    }

}

File: LeeYooseok/BOJ_오늘의 문제/2024/2024-02/2024-02-25/6236 용돈 관리.py
N, M = map(int, input().split())
plan_list = list()
for _ in range(N):
    plan_list.append(int(input()))

left = min(plan_list)
right = sum(plan_list)

answer = 0
while left <= right:
    mid = (left + right) // 2

    money = mid # 현재 가진 동
    num = 1 # 인출 횟수
    for i in range(N):
        # 가진 돈이 부족하면, 돈 다 집어넣고 다시 인출
        if money < plan_list[i]:
            money = mid
            num += 1
        money -= plan_list[i]

    # 인출 횟수가 M 보다 크거나, 인출한 금액이 하루를 다 살기에 부족한 경우
    if num > M or mid < max(plan_list):
        left = mid + 1
    else:
        right = mid - 1
        answer = mid

print(answer)